include/alpaka/core/ThreadPool.hpp

```diff
@@ -65,15 +42,45 @@ namespace alpaka::core
                 std::unique_lock<std::mutex> lock{m_mutex};
                 m_tasks.emplace(std::move(newTask));
             }
+            {
+                std::unique_lock<std::mutex> lock_c{c_mutex};
+                if(m_busyThreads == m_threads.size())
+                {
+                    m_threads.emplace_back(std::thread(
+                        [this]
+                        {
+                            taskType task;
+                            while(true)
+                            {
+                                {
+                                    std::unique_lock<std::mutex> lock{m_mutex};
+                                    m_cond.wait(lock, [this] { return m_stop || !m_tasks.empty(); });
+
+                                    if(m_stop && m_tasks.empty())
+                                        break;
+
+                                    task = std::move(m_tasks.front());
+                                    m_tasks.pop();
+                                }
+
+                                task();
+                                m_busyThreads--;
+                            }
+                        }));
+                }
+            }
             m_cond.notify_one();
+            m_busyThreads++;
+
             return f;
         }
 
     private:
         std::vector<std::thread> m_threads;
         std::condition_variable m_cond;
-        std::mutex m_mutex;
-        bool m_stop = false;
+        std::mutex m_mutex, c_mutex;
+        bool m_stop{false};
         std::queue<taskType> m_tasks;
+        std::atomic<unsigned int> m_busyThreads{0};
```

Comment:
Since this value can become negative when a newly created thread already finishes a task before you reach `m_busyThreads++`, I am inclined to want a signed integral here.